### PR TITLE
Proof of Concept : Using AVX512 masked operations

### PR DESCRIPTION
**Pull request by jfalcou**
This PR is a PoC of how I think we should handle the masked AVX512 operations.
By amending mask_ops to take a callable containing the masked intrinsic, we are able to factor a lot of code around the ignore_*

I am awaiting comments more than validation on this one

### CODE REVIEW

File: include/eve/conditional.hpp [hunk 1]
```diff
@@ -364,12 +353,26 @@ namespace eve
 
     template<typename V> EVE_FORCEINLINE auto else_(V v) const  {  return or_(*this,v);  }
 
-    template<typename T> EVE_FORCEINLINE auto mask(eve::as_<T> const&) const
+    template<typename T> EVE_FORCEINLINE as_logical_t<T> mask(eve::as_<T> const&) const
```
Comment (DenisYaroshevskiy) [inline include/eve/conditional.hpp:356]: Just forward to keep last?

File: include/eve/conditional.hpp [hunk 2]
```diff
@@ -549,13 +560,29 @@ namespace eve
...
+        return typename type::storage_type{mask};
+      }
+      else
+      {
+        using i_t = as_arithmetic_t<as_integer_t<T>>;
+        auto const i = detail::linear_ramp(eve::as_<i_t>());
+        auto const m = (i >= first_count_) && (i < (cardinal_v<T>-last_count_));
+
+        return bit_cast(m, as_<type>());
+      }
```
Comment (DenisYaroshevskiy) [inline include/eve/conditional.hpp:585]: Same here, just forward to the other one?

File: include/eve/detail/function/simd/common/conditional.hpp [hunk 3]
```diff
@@ -52,4 +52,40 @@ namespace eve::detail
       }
     }
   }
+
+  template<conditional_expr C, typename BaseOp, typename MaskOp, typename Arg0, typename... Args>
+  EVE_FORCEINLINE auto mask_op( x86_512_ const&, C const& c
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/conditional.hpp:57]: I don't think this is still needed, is it?

File: include/eve/module/real/core/function/regular/simd/x86/abs.hpp [hunk 4]
```diff
@@ -17,38 +17,85 @@
...
-    {
-          if constexpr( cat == category::int32x8  ) return _mm256_abs_epi32(v);
-    else  if constexpr( cat == category::int16x16 ) return _mm256_abs_epi16(v);
-    else  if constexpr( cat == category::int8x32  ) return _mm256_abs_epi8(v);
-    }
-    else                                            return aggregate(eve::abs, v);
+          if constexpr(c && category::unsigned_  )  return v;
+    else  if constexpr(c == category::float32x16 )  return _mm512_abs_ps(v);
+    else  if constexpr(c == category::float64x8  )  return _mm512_abs_pd(v);
+    else  if constexpr(c && category::float_     )  return bit_notand(mzero(as(v)), v);
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/abs.hpp:27]: What's up with zeroes at the beginning?

File: include/eve/module/real/core/function/regular/simd/x86/abs.hpp [hunk 5]
```diff
@@ -17,38 +17,85 @@
...
+        else  if constexpr(c == category::int8x32   ) return _mm256_mask_abs_epi8 (src,m,a0);
+        else  if constexpr(c == category::int8x16   ) return _mm_mask_abs_epi8    (src,m,a0);
+      };
+
+      return mask_op( x86_512_{}, cx, eve::abs, m_abs, v );
+    }
+    else
+    {
+      return mask_op( cpu_{}, cx, eve::abs, v);
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/abs.hpp:99]: I thought mask_op won't be a thing, or did I misunedrstand?

File: include/eve/conditional.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/conditional.hpp:356]: I was pondering doing so

File: include/eve/module/real/core/function/regular/simd/x86/abs.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/simd/x86/abs.hpp:27]: mzero is 0x80000000 so notand with a float sets the bit of sign to zero in a single instruction

File: include/eve/module/real/core/function/regular/simd/x86/abs.hpp (hunk 5 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/simd/x86/abs.hpp:99]: It'll, this PR is still incomplete

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**